rfcs/0008-storage-handler.md

```diff
@@ -0,0 +1,338 @@
+
+# Motivation
+
+[motivation]: #motivation
+
+The current approach for reading and writing storage slots/variables has multiple flaws, despite being quite ergonomic:
+
+1. There is currently no way to reason about the location of a particular storage variable, which makes passing "references" to storage variables impossible. One workaround is manipulating and reading storage slots manually via `std::storage::store` and `std::storage::get`, but this approach is quite unsafe. Alternatively, one could declare a `trait` in a library and implement it in a contract and have its methods access storage variables. Those methods are then used in the library to access storage indirectly. This does not solve the full problem and feels more like a workaround than a real solution.
+1. Dynamic storage types, such as `StorageMap` and `StorageVec`, currently require a "hacky" compiler intrinsic called `__get_storage_key` which is not very well defined and is hard to use. Introducing the concept of a `StorageKey` will helps us completely remove this intrinsic.
+1. The current implementation of dynamic storage types prevents th-em from being used as struct fields or as type parameters of other storage types (e.g. `StorageVec<StorageVec<u64>>`).
+
+# Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+A `StrorageKey` is a thin wrapper type, in the standard library, around a `b256` key that describes a storage slot:
+
+```rust
+pub struct StorageKey<T> {
+    key: b256,
+}
+```
+
+The type `T` describes the type of the data stored at `key`. Since each storage slot is 64 bytes long, some types cannot fit in a single storage slot. In that case, the data is stored across multiple consecutive storage slots starting at `key`. More on this in [Storage Slot Assignment](#storage-slot-assignment).
+
+Reading and writing the data stored at `key` can be accomplished using the `read` and `write` methods below:
+
+```rust
+impl<T> StorageKey<T> {
+    #[storage(read)]
+    pub fn read(self) -> Option<T> {
+        std::storage::get(self.key)
+    }
+
+    #[storage(write)]
+    pub fn write(self, other: T) {
+        std::storage::store(self.key, other);
+    }
+}
+```
+
+Notice that `read` returns an `Option` because it is possible that a particular storage slot is not valid, i.e. has not been written before, in which case `None` is returned. If the type `T` spans multiple storage slots, then the method `read` returns `None` if at least one the storage slots that `T` spans is invalid.
+
+With the `StorageKey` type available, this RFC proposes a redefinition of the meaning of the expression `storage.<var>.<field>...` to return a `StorageKey` "pointing" to the actual data instead of returning the data itself. The RFC also proposes removing the "reassignment" statement `storage.<var>.<field>.. = ..` from the language.
```

Comment:
What's the reason for removing the reassignment syntax? Why is it problematic to keep it?

If we're keeping the assignment syntax for reading the data, it seems surprising that it won't work for writing. If I can do:

```
let x_key: StorageKey<u64> = storage.s.x
```

I'd expect to be able to do:

```
storage.s.x = 42
```

Reply:
> If I can do:
> 
> let x_key: StorageKey<u64> = storage.s.x
> 
> I'd expect to be able to do:
> 
> storage.s.x = 42

The key difference is that under this proposal, in both of these cases, the types are totally different. E.g. in former, you assign a storage key to a variable of type `StorageKey<u64>`, but in the latter, you attempt to assign an integer to a storage key.

I think the removal of the assignment syntax is also useful for more clearly distinguishing the cost involved in storage interaction vs regular variable/field assignment, e.g. assigning the following two pieces of code have vastly different costs, but it's not at all obvious to a new user reading sway code:
```rust
my_struct.foo.bar = 42;
```
```rust
storage.foo.bar = 42;
```